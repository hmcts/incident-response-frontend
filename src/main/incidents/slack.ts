import config = require('config');

const { App } = require('@slack/bolt');

export const app = new App({
  signingSecret: config.get('secrets.response.slack-signing-secret'),
  token: config.get('secrets.response.slack-bot-token'),
});
