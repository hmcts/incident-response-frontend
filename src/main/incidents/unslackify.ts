import { UsersProfileGetResponse } from '@slack/web-api/dist/response';
import { Profile } from '@slack/web-api/dist/response/UsersProfileGetResponse';

import { app } from './slack';

const { Logger } = require('@hmcts/nodejs-logging');
const NodeCache = require('node-cache');

const logger = Logger.getLogger('incidents/unslackify');

const myCache = new NodeCache({
  stdTTL: 3600,
});

export async function userIdToDisplayName(text: string): Promise<string> {
  const cachedValue = myCache.get(text);
  if (cachedValue) {
    return cachedValue;
  }

  try {
    const user: UsersProfileGetResponse = (
      await app.client.users.info({
        user: text,
      })
    ).user;

    if (!user.profile) {
      return text;
    }

    const name = convertProfileToName(user.profile);
    if (name) {
      myCache.set(text, `@${name}`);
      return `@${name}`;
    }
  } catch (err) {
    logger.error(err);
  }
  return text;
}

/**
 * Users may have a display name set or may not.
 * Display name is normally better than real name, so we prefer that but fallback to real name.
 */
function convertProfileToName(profile: Profile): string | undefined {
  let name = profile.display_name_normalized;
  if (!name) {
    name = profile.real_name_normalized;
  }
  return name;
}

export function userReferenceToId(text: string): string {
  const result = /&lt;@(U[A-Z\d]+)&gt;/.exec(text);
  return result ? result[1] : text;
}

function italics(text: string) {
  const regex = /(?:^_| _|\r\n_|\r_|\n_)(.+)_/g;
  return modifyText(regex, text, matches => {
    return `<span class="hmcts-!-font-style-italic">${matches[0]}</span>`;
  });
}

function bold(text: string) {
  const regex = /(?:^\*| \*|\r\n\*|\r\*|\n\*)(.+)\*/g;
  return modifyText(regex, text, matches => {
    return `<span class="govuk-!-font-weight-bold">${matches[0]}</span>`;
  });
}

function linkWithText(text: string) {
  const regex = /&lt;(https?:\/\/.+)\|(.+)&gt;/g;
  return modifyText(regex, text, matches => {
    return `<a class="govuk-link" href="${matches[0]}">${matches[1]}</a>`;
  });
}

function plainLink(text: string) {
  const regex = /&lt;(https?:\/\/.+)&gt;/g;
  return modifyText(regex, text, matches => {
    return `<a class="govuk-link" href="${matches[0]}">${matches[0]}</a>`;
  });
}

function multiLine(unslackifiedText: string) {
  return unslackifiedText.replace(/\r\n|\r|\n/g, '<br>');
}

async function userIdToDisplayNameConversion(text: string) {
  const regex = /(&lt;@U[A-Z\d]+&gt;)/g;

  return modifyTextAsync(regex, text, text, async matches => {
    const userId = userReferenceToId(matches[0]);
    return userIdToDisplayName(userId);
  });
}

function strikethrough(text: string) {
  const regex = /(?:^~| ~|\r\n~|\r~|\n~)(.+)~/g;
  return modifyText(regex, text, matches => {
    return `<span class="hmcts-!-text-decoration-line-through">${matches[0]}</span>`;
  });
}

function inlineCode(text: string) {
  const regex = /`([^`]+)`/g;
  return modifyText(regex, text, matches => {
    return `<code>${matches[0]}</code>`;
  });
}

function codeBlock(text: string) {
  const regex = /```([^`]+)```/g;

  return modifyText(regex, text, matches => {
    return `<pre><code>${matches[0]}</code></pre>`;
  });
}

function modifyText(regex: RegExp, text: string, replacementText: (matches: string[]) => string) {
  const matches = matchRegex(regex, text);

  let modifiedText = text;
  matches.forEach(theMatch => {
    modifiedText = modifiedText.replace(theMatch.original, replacementText(theMatch.matches));
  });
  return modifiedText;
}

function matchRegex(regex: RegExp, text: string) {
  const matches = [];
  let match = regex.exec(text);
  while (match !== null) {
    matches.push({
      original: match[0],
      matches: [match[1], match[2]],
    });
    match = regex.exec(text);
  }
  return matches;
}

async function modifyTextAsync(
  regex: RegExp,
  text: string,
  unslackifiedText: string,
  replacementTextFn: (matches: string[]) => Promise<string>
) {
  const matches = matchRegex(regex, text);

  let modifiedText = unslackifiedText;
  for (const theMatch of matches) {
    const replacementText = await replacementTextFn(theMatch.matches);

    modifiedText = modifiedText.replace(theMatch.original, replacementText);
  }
  return modifiedText;
}

function blockquote(text: string) {
  // blockquote must be at the start of a line, given we don't do line by line searching, we try to find the start
  const regex = /(?:\r\n|\r|\n)&gt; (.+)/g;
  return modifyText(regex, text, matches => {
    return `<blockquote>${matches[0]}</blockquote>`;
  });
}

/**
 * Adapts slacks markdown format to HTML
 *
 * Looks up users IDs and converts them to a display name
 * @see https://slack.com/intl/en-gb/help/articles/202288908-Format-your-messages
 * @param text the text from response API
 */
export async function unslackify(text: string): Promise<string> {
  let unslackifiedText = await userIdToDisplayNameConversion(text);
  unslackifiedText = bold(unslackifiedText);
  unslackifiedText = italics(unslackifiedText);
  unslackifiedText = strikethrough(unslackifiedText);
  unslackifiedText = linkWithText(unslackifiedText);
  unslackifiedText = plainLink(unslackifiedText);
  unslackifiedText = codeBlock(unslackifiedText);
  unslackifiedText = inlineCode(unslackifiedText);
  unslackifiedText = blockquote(unslackifiedText);

  return multiLine(unslackifiedText);
}
