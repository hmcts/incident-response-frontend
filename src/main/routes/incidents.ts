import * as express from 'express';

import { unslackify } from '../incidents/unslackify';
import { date } from '../modules/nunjucks/filters/date';

const { Logger } = require('@hmcts/nodejs-logging');
const logger = Logger.getLogger('routes/incidents');
const got = require('got');

const router = express.Router();

const responseUrl = process.env.RESPONSE_URL || 'http://localhost:8000';

const client = got.extend({
  prefixUrl: `${responseUrl}/core/incidents`,
  headers: {
    Accept: 'application/json',
  },
  responseType: 'json',
});

type SlackAuthor = {
  app_id: string; // always 'slack'
  external_id: string;
  display_name: string;
  full_name: string;
  email: string;
};

type TimelineMetadata = {
  author: SlackAuthor;
  message_ts: string;
  channel_id: string;
};

type TimelineEvent = {
  id: number;
  timestamp: string;
  text: string;
  event_type: string;
  metadata: TimelineMetadata;
  text_ui: string;
};

function compareTimelineEvents(a: TimelineEvent, b: TimelineEvent) {
  return a.timestamp > b.timestamp ? -1 : 1;
}

/* GET home page. */
router.get('/incident/:id', async (req, res, next) => {
  try {
    const response = await client.get(req.params.id);

    if (req.query.json === 'true') {
      return res.json(response.body);
    }

    const timeline = await client.get(`${req.params.id}/timeline/events`);

    const uiTimeline = await Promise.all(
      timeline.body.results
        // skipping incident_lead promotion for now
        .filter((entry: TimelineEvent) => entry.event_type !== 'incident_update')
        .sort(compareTimelineEvents)
        .map(async (entry: TimelineEvent) => {
          const description = await unslackify(entry.text_ui);

          return {
            title: { text: date(entry.timestamp) },
            by: entry.metadata.author.full_name,
            description: { html: `<div class="hmcts-timeline__description">${description}</div>` },
          };
        })
    );

    res.render('incident', { incident: response.body, timeline: uiTimeline });
  } catch (err) {
    logger.error(err);
    next(err);
  }
});

module.exports = router;
