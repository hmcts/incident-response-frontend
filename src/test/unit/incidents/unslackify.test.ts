import '../../mocks/slack';
import { unslackify, userIdToDisplayName, userReferenceToId } from '../../../main/incidents/unslackify';

describe('unslackify', () => {
  test('converts slack-isms to browser suitable', async () => {
    await expect(unslackify('&lt;@UMMJ5Q47M&gt; hello')).resolves.toBe('@I like rice hello');
  });

  describe('userIdToDisplayName', () => {
    test('converts slack user id to display name', async () => {
      await expect(userIdToDisplayName('UMMJ5Q47M')).resolves.toBe('@I like rice');
    });
  });

  describe('userReferenceToId', () => {
    test('converts slack user id reference to ID', async () => {
      expect(userReferenceToId('&lt;@UMMJ5Q47M&gt;')).toBe('UMMJ5Q47M');
    });

    test('returns existing text if not an id reference', async () => {
      expect(userReferenceToId('not an id')).toBe('not an id');
    });
  });

  test('to match snapshot', async () => {
    const unslackifiedResult = await unslackify(`
    *bold1* non bold

*bold2* not bold2

_Test italic_ and non italic

_Test2 italic2_ and non italic

~strike through~ and non

~strike through2~ and non

\`code\` inline code

\`code23\` inline code

\`\`\`multi-line code

hello\`\`\`
\`\`\`multi-line2 code

hello3\`\`\`

&gt; block quote
&gt; on multi lines
&gt; block quote2
&gt; on multi lines2

1. ordered list
2. second item
1. ordered list2
2. second item2

• bulleted list
• second item bulleted list
• bulleted list2
• second item bulleted list2
Link1:
&lt;https://github.com/hmcts/azure-platform-terraform/commit/82f70a946b14db4616b0b53bb61c8fc5a8f4dab4&gt;

link2:
&lt;https://dev.azure.com/hmcts/CNP/_build/results?buildId=247004&amp;view=ms.vss-test-web.build-test-results-tab&gt;

&lt;https://github.com/hmcts/azure-platform-terraform/commit/82f70a946b14db4616b0b53bb61c8fc5a8f4dab4|link with text1&gt;

&lt;https://dev.azure.com/hmcts/CNP/_build/results?buildId=247004&amp;view=ms.vss-test-web.build-test-results-tab|link with text2&gt;

&lt;@U2YKSMJU9&gt; Hi

&lt;@U2YKSMJU9&gt; Bye

&lt;@U03KG7Y9TRA&gt; let’s make sure this works for people other than me`);

    expect(unslackifiedResult).toMatchSnapshot();
  });
});
