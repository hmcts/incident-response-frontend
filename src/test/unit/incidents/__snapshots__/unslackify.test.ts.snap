// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unslackify to match snapshot 1`] = `"<br>   <span class=\\"govuk-!-font-weight-bold\\">bold1</span> non bold<br><span class=\\"govuk-!-font-weight-bold\\">bold2</span> not bold2<br><span class=\\"hmcts-!-font-style-italic\\">Test italic</span> and non italic<br><span class=\\"hmcts-!-font-style-italic\\">Test2 italic2</span> and non italic<br><span class=\\"hmcts-!-text-decoration-line-through\\">strike through</span> and non<br><span class=\\"hmcts-!-text-decoration-line-through\\">strike through2</span> and non<br><br><code>code</code> inline code<br><br><code>code23</code> inline code<br><br><code><pre>multi-line code<br><br>hello</pre></code><br><code><pre>multi-line2 code<br><br>hello3</pre></code><br><blockquote>block quote</blockquote><blockquote>on multi lines</blockquote><blockquote>block quote2</blockquote><blockquote>on multi lines2</blockquote><br><br>1. ordered list<br>2. second item<br>1. ordered list2<br>2. second item2<br><br>• bulleted list<br>• second item bulleted list<br>• bulleted list2<br>• second item bulleted list2<br>Link1:<br><a class=\\"govuk-link\\" href=\\"https://github.com/hmcts/azure-platform-terraform/commit/82f70a946b14db4616b0b53bb61c8fc5a8f4dab4\\">https://github.com/hmcts/azure-platform-terraform/commit/82f70a946b14db4616b0b53bb61c8fc5a8f4dab4</a><br><br>link2:<br><a class=\\"govuk-link\\" href=\\"https://dev.azure.com/hmcts/CNP/_build/results?buildId=247004&amp;view=ms.vss-test-web.build-test-results-tab\\">https://dev.azure.com/hmcts/CNP/_build/results?buildId=247004&amp;view=ms.vss-test-web.build-test-results-tab</a><br><br><a class=\\"govuk-link\\" href=\\"https://github.com/hmcts/azure-platform-terraform/commit/82f70a946b14db4616b0b53bb61c8fc5a8f4dab4\\">link with text1</a><br><br><a class=\\"govuk-link\\" href=\\"https://dev.azure.com/hmcts/CNP/_build/results?buildId=247004&amp;view=ms.vss-test-web.build-test-results-tab\\">link with text2</a><br><br>@I like rice Hi<br><br>@I like rice Bye<br><br>@I like rice let’s make sure this works for people other than me"`;
